pkg/aggregator/apiserver/handler_proxy.go

```diff
@@ -0,0 +1,185 @@
+package apiserver
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"sync/atomic"
+
+	data "github.com/grafana/grafana-plugin-sdk-go/experimental/apis/data/v0alpha1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apiserver/pkg/endpoints/handlers/responsewriters"
+	endpointmetrics "k8s.io/apiserver/pkg/endpoints/metrics"
+	genericapirequest "k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/klog/v2"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	aggregationv0alpha1 "github.com/grafana/grafana/pkg/aggregator/apis/aggregation/v0alpha1"
+	query "github.com/grafana/grafana/pkg/apis/query/v0alpha1"
+	"github.com/grafana/grafana/pkg/plugins"
+	"github.com/grafana/grafana/pkg/web"
+)
+
+// proxyHandler provides a http.Handler which will proxy traffic to a plugin client.
+type proxyHandler struct {
+	localDelegate         http.Handler
+	client                plugins.Client
+	pluginContextProvider PluginContextProvider
+	handlingInfo          atomic.Value
+}
+
+type proxyHandlingInfo struct {
+	name    string
+	handler *pluginHandler
+}
+
+func proxyError(w http.ResponseWriter, req *http.Request, error string, code int) {
+	http.Error(w, error, code)
+
+	ctx := req.Context()
+	info, ok := genericapirequest.RequestInfoFrom(ctx)
+	if !ok {
+		klog.Warning("no RequestInfo found in the context")
+		return
+	}
+	endpointmetrics.RecordRequestTermination(req, info, "grafana-aggregator", code)
+}
+
+func (r *proxyHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	value := r.handlingInfo.Load()
+	if value == nil {
+		r.localDelegate.ServeHTTP(w, req)
+		return
+	}
+	handlingInfo := value.(proxyHandlingInfo)
+
+	_, ok := genericapirequest.UserFrom(req.Context())
+	if !ok {
+		proxyError(w, req, "missing user", http.StatusInternalServerError)
+		return
+	}
+
+	handlingInfo.handler.ServeHTTP(w, req)
+}
+
+// these methods provide locked access to fields
```

Comment:
Where is the logic of such locking? I see that handlingInfo is an atomic.

Reply:
yeah, this can get away without more complex locking logic because the handlingInfo is only updated by a single controller worker